windows/hidapi_descriptor_reconstruct.h

```diff
@@ -119,9 +119,9 @@ typedef struct hid_pp_link_collection_node_ {
 	USHORT NumberOfChildren;
 	USHORT NextSibling;
 	USHORT FirstChild;
-	ULONG  CollectionType : 8;
-	ULONG  IsAlias : 1;
-	ULONG  Reserved : 23;
+	UINT  CollectionType : 8;
+	UINT  IsAlias : 1;
+	UINT  Reserved : 23;
```

Comment:
revert - this may change binary compatibility
any particular reason for this change?

Reply:
I was very unsure about this one. I'll just drop it. [UINT](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-dtyp/52ddd4c3-55b9-4e03-8287-5392aac0627f) and [ULONG](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-dtyp/32862b84-f6e6-40f9-85ca-c4faf985b822)  should be the same, as `int` and `long` are the same. But the only `int` is covered by the standard (+ implementation defined, of course). When using `ULONG` I should probably also use `%lu` instead of `%u`.

Reply:
this is a part of reverce-engineered implementation of internal HID structures
unless there is an explicit statement in MS documentation that ULONG is an alias for UINT in all cases - we can't change it

Reply:
IT also allows to revert https://github.com/libusb/hidapi/commit/98ab22a33d326e5f3b139d16fd0859fb811c548a ( Suppress Visual C++ warning C4214: nonstandard extension used: bit field types other than int) ->  https://github.com/cmorty/hidapi/actions/runs/6891832863

Reply:
my last statement stands

Reply:
Yea, but it's a bit field of under 32bit. Thus the binary representation is covered by the number of bits of the field, not the type. 

As all bit fields are all smaller than 32, and I think that we can agree, that there is no 16-bit Windows around, `int` and `long` should always be at least 32 bit. `long` might actually get larger, and that is likely to be the bigger risk of braking binary compatibility.

Reply:
For what it's worth https://learn.microsoft.com/en-us/windows/win32/winprog/windows-data-types?redirectedfrom=MSDN stats that they are both unsigned 32bit. But, I you insist I'll drop the commit.

Reply:
> Yea, but it's a bit field of under 32bit. Thus the binary representation is covered by the number of bits of the field, not the type.

But it changes the _size of the struct_ it is contained in thus it breaks binary compatibility of the array of such structs (if any).

Reply:
@JoergAtGithub do you have an oppinion on this change? You probably knows this better.

Reply:
...no it doesn't.  The type of a bit-field does not change it's size. Only the number behind the colon is relevant. And that the type is large enough to hold the bit-field.

> Bit-fields
>
> Declares a member with explicit width, in bits. Adjacent bit-field members may be packed to share and straddle the individual bytes. 

https://en.cppreference.com/w/c/language/bit_field

And the relevant sentences from the standard:

> 4)
> The expression that specifies the width of a bit-field shall be an integer constant
> expression with a nonnegative value that does not exceed the width of an object of the
> type that would be specified were the colon and expression omitted.  If the value is
> zero, the declaration shall have no declarator.
> 
> 11)
> An implementation may allocate any addressable storage unit large enough to hold a bit-
> field. If enough space remains, a bit-field that immediately follows another bit-field in a
> structure shall be packed into adjacent bits of the same unit. If insufficient space remains,
> whether a bit-field that does not fit is put into the next unit or overlaps adjacent units is
> implementation-defined. The order of allocation of bit-fields within a unit (high-order to
> low-order or low-order to high-order) is implementation-defined. The alignment of the
> addressable storage unit is unspecified.

https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1548.pdf - §6.7.2.1


According to the standard compilers may actually pad bit-fields. I can't imagine though, that some compiler will ever create padded bit fields, because it will break so much stuff.

Reply:
> The type of a bit-field does not change it's size.

I'm refering to the cases like this: 
https://godbolt.org/z/czEzjzKrG

Reply:
We shouldn't change these, as these are the same types as in the public headers from Microsoft. Using exactly the same is the safest way to keep the binary compatibility.
And binary compatibility of these structs is not covered by our unit tests.

Reply:
https://godbolt.org/z/5WhYsMW14

Reply:
lets revert then

Reply:
Ok, I learned something and hate C/C++ even more. :-( https://godbolt.org/z/9on94frPz
Based on that using UINT seems even more reasonable, but MS's public headers are the killer argument here. I'll drop the commit.

Reply:
OK. I reassured my pedantic me by adding a `static_assert`. I hope that solution is fine with you.